src/device.ts

```diff
@@ -10,61 +10,83 @@
 
 'use strict';
 
-const Action = require('./action');
-const Ajv = require('ajv');
+import { Action, ActionDescription } from './action';
+import Ajv from 'ajv';
+import { Adapter } from './adapter';
+import { Property, PropertyDescription } from './property';
+import { Event, EventDescription } from './event';
 const ajv = new Ajv();
 
-class Device {
-  constructor(adapter, id) {
+export interface DeviceDescription {
+  id: string,
+  title: string,
+  '@context': string,
+  '@type': string[],
+  description: string,
+  properties: { [key: string]: PropertyDescription },
+  actions: { [key: string]: ActionDescription },
+  events: { [key: string]: EventDescription },
+  links: Link[],
+  baseHref?: string,
+  pin: {
+    required: boolean,
+    pattern?: string,
+  },
+  credentialsRequired: boolean,
+}
+
+export interface Link {
+  rel: string,
+  href: string
+}
+
+export class Device {
+  private '@context' = 'https://iot.mozilla.org/schemas';
+  private '@type': string[] = [];
+  private name: string = '';
+  private title = '';
+  private description = '';
+  private properties = new Map<string, Property>();
+  private actions = new Map<string, ActionDescription>();
+  private events = new Map<string, EventDescription>();
+  private links: Link[] = [];
+  private baseHref?: string;
+  private pinRequired = false;
+  private pinPattern?: string;
+  private credentialsRequired = false;
+
+  constructor(private adapter: Adapter, private id: string) {
     if (typeof id !== 'string') {
-      id = id.toString();
+      id = (<any>id).toString();
     }
-
-    this.adapter = adapter;
-    this.id = id;
-    this['@context'] = 'https://iot.mozilla.org/schemas';
-    this['@type'] = [];
-    this.title = '';
-    this.description = '';
-    this.properties = new Map();
-    this.actions = new Map();
-    this.events = new Map();
-    this.links = [];
-    this.baseHref = null;
-    this.pinRequired = false;
-    this.pinPattern = null;
-    this.credentialsRequired = false;
   }
 
-  asDict() {
-    const properties = {};
-    this.properties.forEach((property, propertyName) => {
-      properties[propertyName] = property.asDict();
-    });
-
-    const actions = {};
-    this.actions.forEach((metadata, actionName) => {
-      actions[actionName] = Object.assign({}, metadata);
+  mapToDict<V>(map: Map<string, V>) {
+    const dict: { [key: string]: V } = {};
+    map.forEach((property, propertyName) => {
+      dict[propertyName] = property;
     });
+    return dict;
+  }
 
-    const events = {};
-    this.events.forEach((metadata, eventName) => {
-      events[eventName] = Object.assign({}, metadata);
+  mapToDictF<V>(map: Map<string, { asDict: () => V }>) {
```

Comment:
What "F" in name stands for? :-)

Reply:
`Function` :grimacing: 
If you have a better name for the function go ahead :smile:

Reply:
`mapToDictWithFunction`? I am cool with current one, BTW.

Reply:
Done